### ISSUE TITLE
curl overrides are discarded when pulled in via fetchpatch, but overrideAttrs are not

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#66499]
I have this blocker problem currently for `musl` builds in [static-haskell-nix](https://github.com/nh2/static-haskell-nix/):

When I override `curl` in an overlay, while also adding a `fetchpatch` to the proper `libdrm/default.nix`, then the the fetchpatch causes `curl` to build and fail with linker error.

That is because while it does build the curl from my overlay, it seems to ignore  what I've given in `.override` (e.g. I give `gssSupport = false`, but inside e.g. `configurePhase` it's still `true` again).

I basically do in my overlay

```nix
{
  curl = (previous.curl.override { gssSupport = false; }).overrideAttrs (old: {
    makeFlags = ["V=1"];
  });
}
```

and observe that the `V=1` is in effect, but the `gssSupport = false` is not.

(I see that in `nix show-derivation` on the `.drv`.)

What's going on here?

---

cc from IRC: @matthewbauer @Ericson2314 @dtzWill @cleverca22 

##### Environment

commit bc94dcf500286495e3c478a9f9322debc94c4304

**Comment by nh2** [comment]
To repro this:

```bash
git clone --recursive https://github.com/nh2/static-haskell-nix.git -b nixpkgs-66499-problem-repro-curl-override-discarded
cd static-haskell-nix
NIX_PATH=. nix-build --no-link survey/default.nix --arg disableOptimization true -A pkgs.libdrm
```

(this is the branch in https://github.com/nh2/static-haskell-nix/pull/45)

---

It prints

```
...
trace: gssSupport true
these derivations will be built:
  /nix/store/bkk1f6ls0rm6xz4l2fyy9gxlzlfafy7x-curl-7.64.1.drv
  /nix/store/2hcg7wnzyqamdy68y1rhcpqf0gap1jl9-musl-ioctl.patch.drv
  /nix/store/150vlxnw31ida8wr45fknywjm8kkn7hs-libdrm-2.4.97.drv
building '/nix/store/bkk1f6ls0rm6xz4l2fyy9gxlzlfafy7x-curl-7.64.1.drv'...
...
configuring

gssSupport true

exiting
builder for '/nix/store/bkk1f6ls0rm6xz4l2fyy9gxlzlfafy7x-curl-7.64.1.drv' failed with exit code 1
```

And

```
% nix show-derivation /nix/store/bkk1f6ls0rm6xz4l2fyy9gxlzlfafy7x-curl-7.64.1.drv | grep -E 'configureFlags|makeFlags'
      "configureFlags": "--without-ca-bundle --without-ca-path --with-ca-fallback --disable-manual --with-ssl=/nix/store/c7ax9w05g2vp36wm0j695fbh89dma2rq-openssl-1.0.2r-dev --without-gnutls --with-libssh2=/nix/store/a6crgyyg9v9splnidqmz2530q888cf6l-libssh2-1.8.2-dev --disable-ldap --disable-ldaps --without-libidn --without-brotli --with-gssapi=/nix/store/q82pls5sglxi3bi49r7m736kz01pqmkc-libkrb5-1.17-dev",
      "makeFlags": "curl_LDFLAGS=-all-static V=1",
```

shows that my `makeFlags` from [here](https://github.com/nh2/static-haskell-nix/pull/45/commits/6bc84ad1a27b4340902736899d48c9dc29260cad#diff-5a873d92c00133aa8d3f0a425a93aac2R630) made it in but `--with-gssapi` is given even though I give `gssSupport = false` [here](https://github.com/nh2/static-haskell-nix/pull/45/commits/6bc84ad1a27b4340902736899d48c9dc29260cad#diff-5a873d92c00133aa8d3f0a425a93aac2R587).

**Comment by nh2** [comment]
Is there some magic somewhere that makes that `fetchpatch -> fetchurl -> curl` discards `curl.override` in overlays but retains `curl.overrideAttrs`?

**Comment by nh2** [comment]
Some info from @Infinisil:

> Not sure, but curl is a bit special in that a modified version is used for `pkgs.fetchurl`
> https://github.com/NixOS/nixpkgs/blob/ac2eb3151df879602c58280d09e9c205fcff8534/pkgs/top-level/all-packages.nix#L278
> There `buildPackages.curl` is used with `gssSupport` overwritten

I confirmed:

> that place certainly seems to be very involved in the problem, if I set `gssSupport = false;` then it's false in my build

So indeed that place just overrides my `.override` but not my `overrideAttrs`.

How can I make my `.override` take precedence?

**Comment by Infinisil** [comment]
Maybe something like
```nix
self: super: {
  fetchurl = super.fetchurl.override (old: {
    curl = old.curl.override {
      gssSupport = false;
      # ...
    };
  });
}`

**Comment by nh2** [comment]
I should also mention why I actually want to do that:

In my overlay, I want to have `libcurl` and `libkrb5` have static libraries (`.a` files).

So I do

```nix
{
    krb5 = previous.krb5.override {
      # Note krb5 does not support building both static and shared at the same time.
      # That means *anything* on top of this overlay trying to link krb5
      # dynamically from this overlay will fail with linker errors.
      staticOnly = true;
    };

    # Disable gss support, because that requires `krb5`,
    # which (as mentioned above) is a library that cannot build both
    # .a and .so files in its build system.
    # That means that if we enable it, we can no longer build the
    # dynamically-linked `curl` binary from this overlay.
    # But `fetchurl`, `fetchpatch` and so on by default depend on dynamically
    # linked `curl` binaries.
    curl = (previous.curl.override { gssSupport = false; }).overrideAttrs (old: {
      dontDisableStatic = true;
    });
}
```

See the dilemma? Having `krb5` `.a` files means there will be no `.so` files for it, which means `curl` can't find krb5 `.so` files, so I must disable that feature for `curl`.

**Comment by nh2** [comment]
> ```nix
> fetchurl = super.fetchurl.override (old: {
> ```

That doesn't work, because `super.fetchurl` is a function taking `{ url, ... }`, not a derivation, so it doesn't have an `.override` attribute.

It seems `fetchurl` can't be overridden.

**Comment by nh2** [comment]
@Infinisil said:

> Try changing nixpkgs `fetchurl = import ../build-support/fetchurl` to `fetchurl = callPackage ...

Indeed `callPackage` seems to make it work.

me:

> Should that be changed this way?

@Infinisil siad:

> Yeah I'd say so
> Seeing the eval stats for that will be interesting

**Comment by nh2** [comment]
I found the fix in #66506.
